File: frontend/src/services/leaveService.js
/**
 * Leave service — connects to the backend leave API endpoints.
 * Leave balances remain frontend-only (no backend endpoint).
 */

import { apiClient } from './apiClient';

/**
 * Transform a backend LeaveOut to the frontend display format.
 */
const transformLeave = (leave) => ({
  id: leave.id,
  employee_id: leave.employee_id, // numeric employee ID from backend
  leave_type: leave.leave_type,
  start_date: leave.start_date,
  end_date: leave.end_date,
  remarks: leave.remarks || '',
  status: leave.status,
  approved_by: leave.approved_by,
  approval_comment: leave.approval_comment,
  created_at: leave.created_at,
  updated_at: leave.updated_at,
});

export const leaveService = {
  /**
   * Submit a leave request.
   * POST /api/leave → LeaveOut
   */
  create: async (data) => {
    const body = {
      leave_type: data.leave_type,
      start_date: data.start_date,
      end_date: data.end_date,
      remarks: data.remarks || null,
    };
    const result = await apiClient.post('/leave', body);
    return transformLeave(result);
  },

  /**
   * List leave requests.
   * GET /api/leave?employee_id=X&status=Y → list[LeaveOut]
   * Admin sees all; Employee sees own.
   */
  getAll: async (employeeId = null, status = null) => {
    const params = {};
    if (employeeId) params.employee_id = employeeId;
    if (status) params.status = status;
    const data = await apiClient.get('/leave', params);
    return data.map(transformLeave);
  },

  /**
   * Get a single leave request by ID.
   * GET /api/leave/{id} → LeaveOut
   */
  getById: async (id) => {
    const data = await apiClient.get(`/leave/${id}`);
    return transformLeave(data);
  },

  /**
   * [Admin] Approve or reject a leave request.
   * PATCH /api/leave/{id}/decide → LeaveOut
   */
  decide: async (id, status, approvalComment = null) => {
    const body = {
      status: status,
    };
    if (approvalComment) {
      body.approval_comment = approvalComment;
    }
    const data = await apiClient.patch(`/leave/${id}/decide`, body);
    return transformLeave(data);
  },
};

File: backend/app/services/leave_service.py
"""Leave request service — creation, listing, and Admin approval workflow."""

import logging
from datetime import date

from sqlalchemy import and_, select
from sqlalchemy.ext.asyncio import AsyncSession

from app.core.exceptions import ConflictError, ForbiddenError, NotFoundError
from app.models.employee import Employee
from app.models.leave import LeaveRequest, LeaveStatus
from app.models.user import User, UserRole
from app.schemas.leave import LeaveCreate, LeaveDecision

logger = logging.getLogger(__name__)


async def create_leave_request(
    db: AsyncSession,
    employee: Employee,
    data: LeaveCreate,
) -> LeaveRequest:
    """Employee submits a leave request."""
    # Check for overlapping approved/pending leaves
    overlap = await db.execute(
        select(LeaveRequest).where(
            and_(
                LeaveRequest.employee_id == employee.id,
                LeaveRequest.status != LeaveStatus.REJECTED,
                LeaveRequest.start_date <= data.end_date,
                LeaveRequest.end_date >= data.start_date,
            )
        )
    )
    if overlap.scalar_one_or_none():
        raise ConflictError(
            "A leave request already exists that overlaps with the requested dates"
        )

    req = LeaveRequest(
        employee_id=employee.id,
        leave_type=data.leave_type,
        start_date=data.start_date,
        end_date=data.end_date,
        remarks=data.remarks,
        status=LeaveStatus.PENDING,
    )
    db.add(req)
    await db.commit()
    await db.refresh(req)
    logger.info(
        "Leave request created  employee_id=%s  leave_id=%s  type=%s  dates=%s→%s",
        employee.id,
        req.id,
        data.leave_type,
        data.start_date,
        data.end_date,
    )
    return req


async def list_leave_requests(
    db: AsyncSession,
    requesting_user: User,
    employee_id: int | None = None,
    status: LeaveStatus | None = None,
) -> list[LeaveRequest]:
    """
    Admin: list all or filter by employee/status.
    Employee: list their own requests only.
    """
    stmt = select(LeaveRequest)

    if requesting_user.role != UserRole.ADMIN:
        # Resolve the employee record for this user
        emp_result = await db.execute(
            select(Employee).where(Employee.user_id == requesting_user.id)
        )
        emp = emp_result.scalar_one_or_none()
        if not emp:
            return []
        stmt = stmt.where(LeaveRequest.employee_id == emp.id)
    elif employee_id is not None:
        stmt = stmt.where(LeaveRequest.employee_id == employee_id)

    if status is not None:
        stmt = stmt.where(LeaveRequest.status == status)

    result = await db.execute(stmt.order_by(LeaveRequest.created_at.desc()))
    return list(result.scalars().all())


async def get_leave_request(db: AsyncSession, leave_id: int) -> LeaveRequest:
    result = await db.execute(
        select(LeaveRequest).where(LeaveRequest.id == leave_id)
    )
    req = result.scalar_one_or_none()
    if not req:
        raise NotFoundError(f"Leave request {leave_id} not found")
    return req


async def decide_leave_request(
    db: AsyncSession,
    leave_id: int,
    admin: User,
    decision: LeaveDecision,
) -> LeaveRequest:
    """Admin approves or rejects a leave request."""
    req = await get_leave_request(db, leave_id)

    if req.status != LeaveStatus.PENDING:
        raise ConflictError(
            f"Leave request is already in terminal state '{req.status.value}' — cannot change"
        )

    req.status = decision.status
    req.approved_by = admin.id
    req.approval_comment = decision.approval_comment
    db.add(req)
    await db.commit()
    await db.refresh(req)

    logger.info(
        "Leave request decided  leave_id=%s  status=%s  admin_id=%s",
        leave_id,
        decision.status,
        admin.id,
    )
    return req


async def require_leave_access(
    requesting_user: User,
    leave_request: LeaveRequest,
    db: AsyncSession,
) -> None:
    """Raise ForbiddenError if non-admin accesses another employee's leave."""
    if requesting_user.role == UserRole.ADMIN:
        return
    emp_result = await db.execute(
        select(Employee).where(Employee.user_id == requesting_user.id)
    )
    emp = emp_result.scalar_one_or_none()
    if not emp or emp.id != leave_request.employee_id:
        raise ForbiddenError("You do not have permission to access this leave request")
